### PR TITLE
`.timeout(ms, {cancel: true})` to cancel query after timeout (only implemented for MySQL)

### PR DESCRIPTION
**Pull request by byronmwong**
Taking a stab at addressing #707 

Adds an optional `options` argument to `.timeout()`.

For MySQL, a `KILL threadId` query will get sent. For all other dialects an error will be thrown immediately from the `.timeout()` call.

While writing tests, I discovered some potentially worrying behavior (comment also added in the test). After a timeout, `knex` releases the connection back to the pool, but the connection may not be usable because the original slow query is still being performed. A subsequent query will acquire the connection (still in-use) and hang until the first query finishes.

### CODE REVIEW

File: src/query/builder.js [hunk 1]
```diff
@@ -57,9 +57,16 @@ assign(Builder.prototype, {
     return cloned;
   },
 
-  timeout(ms) {
+  timeout(ms, {cancel} = {}) {
     if(isNumber(ms) && ms > 0) {
       this._timeout = ms;
+      if (cancel) {
```
Comment (wubzz) [inline src/query/builder.js]: Unnecessary nested if-statements here. `if (cancel && this.client.canCancelQuery)`

File: src/query/builder.js [hunk 2]
```diff
@@ -57,9 +57,16 @@ assign(Builder.prototype, {
...
+        if (this.client.canCancelQuery) {
+          this._cancelOnTimeout = true;
+        } else {
+          throw new Error("Query cancelling not supported for this dialect");
```
Comment (wubzz) [inline src/query/builder.js]: Is this really needed here given the default client function? I would rather have the second argument ignored for other dialects and make a note of dialect support in documentation than duplicating code.

File: src/raw.js [hunk 3]
```diff
@@ -30,9 +30,16 @@ assign(Raw.prototype, {
     return this
   },
 
-  timeout(ms) {
+  timeout(ms, {cancel} = {}) {
     if(isNumber(ms) && ms > 0) {
       this._timeout = ms;
+      if (cancel) {
```
Comment (wubzz) [inline src/raw.js]: Unnecessary nested if-statements here. `if (cancel && this.client.canCancelQuery)`

File: src/raw.js [hunk 4]
```diff
@@ -30,9 +30,16 @@ assign(Raw.prototype, {
...
+        if (this.client.canCancelQuery) {
+          this._cancelOnTimeout = true;
+        } else {
+          throw new Error("Query cancelling not supported for this dialect");
```
Comment (wubzz) [inline src/raw.js]: Is this really needed here given the default client function? I would rather have the second argument ignored for other dialects and make a note of dialect support in documentation than duplicating code.

File: test/integration/builder/additional.js [hunk 5]
```diff
@@ -299,7 +299,77 @@ module.exports = function(knex) {
...
+        }
+      };
+
+      if(!testQueries.hasOwnProperty(dialect)) {
+        throw new Error('Missing test query for dialect: ' + dialect);
+      }
+
+      var query = testQueries[dialect]();
+
+      var addTimeout = () => query.timeout(1, {cancel: true})
```
Comment (wubzz) [inline test/integration/builder/additional.js]: Unfortunately we can't es6 in the test suite, it's being run for 0.10 and 0.12 as well.

**Comment by wubzz** [comment]
I understand the issue here and don't mind the implementation. Would be great to support other dialects as well later on.

What I'm curious about is what would happen if the `cancelQuery` function is unable to `acquireConnection` due to the pool being full *(or no resources available)*.

**Comment by byronmwong** [comment]
@wubzz thanks for the quick review!

Good point about `cancelQuery` not being to acquire a connection. I could put a timeout on the cancel part as well and throw a different error, e.g. `Unable to cancel query that hit timeout`. What do you think?

File: src/query/builder.js (hunk 2 above)
Comment (byronmwong) [inline src/query/builder.js]: The default client function won't throw until the query has been sent and already timed out. I felt it'd give a better dev experience if we threw earlier.

File: test/integration/builder/additional.js (hunk 5 above)
Comment (byronmwong) [inline test/integration/builder/additional.js]: :+1: will fix

**Comment by wubzz** [comment]
@byronmwong Yeah some form of fallback might be needed. Perhaps the simplest solution would be to have the promise "resolve" itself after timing out. That way the original expected timeout error would still be thrown.

**Comment by byronmwong** [comment]
@wubzz yep that could work, but then we're silently failing on the cancel part.

If a dev calls `.timeout(ms, {cancel: true})` and receives a `TimeoutError`, they'd expect the original query to be cancelled.

File: test/integration/builder/additional.js [hunk 6]
```diff
@@ -256,9 +256,9 @@ module.exports = function(knex) {
     });
 
 
-    it('.timeout() should throw TimeoutError', function(done) {
+    it('.timeout() should throw TimeoutError', function() {
       var dialect = knex.client.config.dialect;
-      if(dialect === 'sqlite3') { return done(); } //TODO -- No built-in support for sleeps
+      if(dialect === 'sqlite3') { return; } //TODO -- No built-in support for sleeps
```
Comment (jurko-gospodnetic) [inline test/integration/builder/additional.js]: you can make sqlite sleep implicitly by using two separate transactions/connections, one locks a database table, and another attempts to access that table, which will cause it to wait for the lock to be released (5 seconds by default, if I recall correctly)

File: src/query/builder.js (hunk 2 above)
Comment (jurko-gospodnetic) [inline src/query/builder.js]: Would it be possible to make the client-type-agnostic part of knex perform this check by having it provide a wrapper `timeout()` function that then calls this function only if the check passes? That would kill the duplication and still give us the immediate error feedback.

Haven't checked the relevant knex code though, so might me talking out of my ass here. :smile:

File: test/integration/builder/additional.js (hunk 6 above)
Comment (byronmwong) [inline test/integration/builder/additional.js]: ohh that's a good idea! @jurko-gospodnetic @wubzz is this a blocker on merging?

**Comment by byronmwong** [comment]
@wubzz: added a timeout on acquiring connection for cancelling queries and got tests to pass. ready for another look!

**Comment by wubzz** [comment]
Looks good to me. Given that the original timeout test was written by me, I don't think the responsibility of adding sqlite support in the test falls on you, so I don't think it's a blocker for this PR.

A quick search shows that `postgres`, `maria` and `oracle` should have support for this action as well. Is this something to look into in this PR, or would you rather a separate issue?

**Comment by byronmwong** [comment]
Turns out the implementation for `maria` is the same as `mysql`, so I've
handled that. I'd prefer a separate issue for `postgres` and `oracle`
support. Thanks!

**Comment by wubzz** [comment]
@byronmwong That's alright with me, I'll make a new issue for postgres/oracle support.

One last thing I almost forgot before I merge this, could you update the docs *(index.html)* of `timeout` to reflect this change? Make sure to specify it only applies to mysql/maria *for now*.

**Comment by byronmwong** [comment]
done! no worries, I should've thought to update the docs myself.

**Pull request merged**